test/job/cpulimit/cpulimit.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"runtime"
+
+	"github.com/adalton/teleport-exercise/pkg/cgroup/v1"
+	"github.com/adalton/teleport-exercise/pkg/jobmanager"
+)
+
+func runTest(controllers ...cgroup.Controller) {
+
+	job := jobmanager.NewJob("theOwner", "my-test", controllers,
+		"/usr/bin/stress-ng",
+		"--cpu",
+		fmt.Sprintf("%d", runtime.NumCPU()),
+		"--timeout",
+		"20",
+		"--times",
+	)
+
+	if err := job.Start(); err != nil {
+		panic(err)
+	}
+
+	for output := range job.StderrStream().Stream() {
+		fmt.Print(string(output))
+	}
+	fmt.Printf("\n")
+}
+
+// Sample run:
+//     $ sudo go run test/job/cpulimit/cpulimit.go
+//     Password:
+//     Running CPU test with no cgroup constraints
+//     stress-ng: info:  [1] setting to a 20 second run per stressor
+//     stress-ng: info:  [1] dispatching hogs: 12 cpu
+//     stress-ng: info:  [1] successful run completed in 20.01s
+//     stress-ng: info:  [1] for a 20.01s run time:
+//     stress-ng: info:  [1]     240.09s available CPU time
+//     stress-ng: info:  [1]     239.37s user time   ( 99.70%)
+//     stress-ng: info:  [1]       0.05s system time (  0.02%)
+//     stress-ng: info:  [1]     239.42s total time  ( 99.72%)
+//     stress-ng: info:  [1] load average: 5.83 2.88 2.22
+//
+//     Running CPU test with cgroup constraints at 0.5 CPU
+//     stress-ng: info:  [1] setting to a 20 second run per stressor
+//     stress-ng: info:  [1] dispatching hogs: 12 cpu
+//     stress-ng: info:  [1] successful run completed in 20.13s
+//     stress-ng: info:  [1] for a 20.13s run time:
+//     stress-ng: info:  [1]     241.60s available CPU time
+//     stress-ng: info:  [1]      10.08s user time   (  4.17%)
+//     stress-ng: info:  [1]       0.02s system time (  0.01%)
+//     stress-ng: info:  [1]      10.10s total time  (  4.18%)
+//     stress-ng: info:  [1] load average: 5.47 2.99 2.27
+
+func main() {
+	fmt.Println("Running CPU test with no cgroup constraints")
+	runTest()
+
+	fmt.Println("Running CPU test with cgroup constraints at 0.5 CPU")
+	runTest(cgroup.NewCpuController().SetCpus(0.5))
```

Comment:
The limits tests look alright, but I wish they could 1) assert the results (and fail accordingly) and 2) were `go test`-able, so they could run in automated fashion.

Reply:
I look at these as a different class of test than ones that are `go test`-able in that they're more long-running than a unit test (for example, this test runs in ~40 seconds).  It'd be possible to parse the output and give it a pass/fail, but would require a little logic to know what the the values should be given the number of CPUs that are available.

Also, this depends on an external tool.  Since we're not currently running this in a chroot, the automation environment would have to have that tool installed.

Reply:
I think we should at least try to have the binaries return a failed exit code, so it could be reasonably automated. This means they have more knowledge of what is going on and are capable of checking their output, so to speak.

As for making it `go test`-able, there are a few ways we can go about this: Benchmark tests, testing.Short(), setting a build tag, env var + t.Skip() - whatever seems more appropriate. Even if we don't, a quick test that asserts the happy path of the library would be great to have.

Reply:
That makes sense.  If that's something you'd like to see in the context of this exercise, then that's something I'll be happy to take a pass at.  Otherwise, I'd suggest we consider this a future enhancement.

Reply:
I think this one is worth having, but we can do it as a follow up to the library PR if you prefer. Your call.

Reply:
:+1: 

I've taken a pass at making some of the tests panic on failure.  I'll take a stab at the others, along with reworking these to use the test framework on Wednesday.

Reply:
Done.  I've refactored all the test binaries to use the test framework, and I've added a `inttest` target in the Makefile to kick them off.